lib/puppet/pops/types/p_object_type.rb

```diff
@@ -0,0 +1,539 @@
+    end
+
+    # @api public
+    def ==(o)
+      eql?(o)
+    end
+
+    # Returns the member as a hash suitable as an argument for constructor. Name is excluded
+    # @return [Hash{String=>Object}] the initialization hash
+    # @api private
+    def i12n_hash
+      hash = { KEY_TYPE => @type }
+      hash[KEY_FINAL] = true if @final
+      hash[KEY_OVERRIDE] = true if @override
+      hash[KEY_ANNOTATIONS] = @annotations unless @annotations.nil?
+      hash
+    end
+
+    # @api private
+    def feature_type
+      raise NotImplementedError, "'#{self.class.name}' should implement #feature_type"
+    end
+
+    # @api private
+    def label
+      "#{feature_type} '#{name}'"
+    end
+  end
+
+  # Describes a named Attribute in an Object type
+  # @api public
+  class PAttribute < PAnnotatedMember
+
+    # @return [String,nil] The attribute kind as defined by #TYPE_ATTRIBUTE_KIND, or `nil` to
+    #   indicate that
+    attr_reader :kind
+
+    # @return [Object] the default value or the symbol :default to denote that there is no default value
+    # @api public
+    attr_reader :value
+
+    # @param name [String] The name of the attribute
+    # @param [Hash{String=>Object}] i12n_hash
+    # @option i12n_hash [PAnyType] 'type' The attribute type (required)
+    # @option i12n_hash [Object] 'value' The default value, must be an instanceof the given `type` (optional)
+    # @option i12n_hash [String] 'kind' The attribute kind, matching #TYPE_ATTRIBUTE_KIND
+    # @api public
+    def initialize(name, i12n_hash)
+      super(name, TypeAsserter.assert_instance_of(["initializer for attribute '%s'", name], TYPE_ATTRIBUTE, i12n_hash))
+      @kind = i12n_hash[KEY_KIND]
+      @final = true if @kind == ATTRIBUTE_KIND_CONSTANT # final is implied
+
+      if i12n_hash.include?(KEY_VALUE)
+        if @kind == ATTRIBUTE_KIND_DERIVED || @kind == ATTRIBUTE_KIND_GIVEN_OR_DERIVED
+          raise Puppet::ParseError, "#{label} of kind '#{@kind}' cannot be combined with an attribute value"
+        end
+        @value = TypeAsserter.assert_instance_of(nil, type, i12n_hash[KEY_VALUE]) {"#{label} #{KEY_VALUE}" }
+      else
+        raise Puppet::ParseError, "#{label} of kind 'constant' requires a value" if @kind == ATTRIBUTE_KIND_CONSTANT
+        @value = :default
+      end
+    end
+
+    # @api public
+    def eql?(o)
+      super && @kind == o.kind && @value == o.value
+    end
+
+    # Returns the member as a hash suitable as an argument for constructor. Name is excluded
+    # @return [Hash{String=>Object}] the hash
+    # @api private
+    def i12n_hash
+      hash = super
+      unless @kind.nil?
+        hash[KEY_KIND] = @kind
+        hash.delete(KEY_FINAL) if @kind == ATTRIBUTE_KIND_CONSTANT # final is implied
+      end
+      hash[KEY_VALUE] = @value unless @value == :default
+      hash
+    end
+
+    # @return [Boolean] `true` if a value has been defined for this attribute. The value may be `nil`
+    def value?
+      value != :default
```

Comment:
no, not at all - we agreed that a literal default had no special meaning - whereas earlier it meant a default value of the type specified for the attribute. It is valid to have an attribute that accepts literal default and you may want it to default to that. You cannot use :default to signal "no value" here.